garage/tf/models/gaussian_mlp_model.py

```diff
@@ -0,0 +1,220 @@
+                    output_b_init=b,
+                    name="mean_std_network")
+
+                with tf.variable_scope("mean_network"):
+                    mean_network = mean_std_network[..., :self._output_dim]
+                with tf.variable_scope("std_network"):
+                    std_network = mean_std_network[..., self._output_dim:]
+            else:
+                mean_network = mlp(
+                    input_var=input_var,
+                    output_dim=self._output_dim,
+                    hidden_sizes=self._hidden_sizes,
+                    hidden_nonlinearity=self._hidden_nonlinearity,
+                    output_nonlinearity=self._output_nonlinearity,
+                    name="mean_network")
+
+                if self._adaptive_std:
+                    b = tf.constant_initializer(self._init_std_param)
+                    std_network = mlp(
+                        input_var=input_var,
+                        output_dim=self._output_dim,
+                        hidden_sizes=self._std_hidden_sizes,
+                        hidden_nonlinearity=self._std_hidden_nonlinearity,
+                        output_nonlinearity=self._output_nonlinearity,
+                        output_b_init=b,
+                        name="std_network")
+
+                else:
+                    p = tf.constant_initializer(self._init_std_param)
+                    std_network = parameter(
+                        input_var=input_var,
+                        length=self._output_dim,
+                        initializer=p,
+                        trainable=self._learn_std,
+                        name="std_network")
+
+            mean_var = mean_network
+            std_param_var = std_network
+
+            with tf.variable_scope("std_limits"):
+                if self._min_std_param:
+                    std_param_var = tf.maximum(std_param_var,
+                                               self._min_std_param)
+                if self._max_std_param:
+                    std_param_var = tf.minimum(std_param_var,
+                                               self._max_std_param)
+
+        with tf.variable_scope("std_parameterization"):
+            # build std_var with std parameterization
+            if self._std_parameterization == "exp":
+                std_var = tf.exp(std_param_var)
+            elif self._std_parameterization == "softplus":
+                std_var = tf.log(1. + tf.exp(std_param_var))
+            else:
+                raise NotImplementedError
+
+        dist = tf.contrib.distributions.MultivariateNormalDiag(
+            mean_var, std_var)
+        sample_var = dist.sample(seed=ext.get_seed())
+
+        inputs = {
+            "input_var": input_var,
+        }
+        outputs = {
+            "mean": mean_var,
+            "std": std_var,
+            "std_param": std_param_var,
+            "sample": sample_var,
+        }
+        model_info = {
+            "dist": dist,
```

Comment:
I think we need to think harder about what the outputs are here. Is it just sample()? Or what about just the dist? Mean and std/std_param seem more like infos than outputs.

Should we even have infos? Isn't that a violation of encapsulation? What is the use case for infos? If your model provides a symbolic output as part of its interface, that is by definition an output.

Reply:
i think dist is an output. This model is a Gaussian with an MLP input. How is the dist not the output?

On the contrary, I think mean, std, and std_param are infos, not outputs. I'm not sure sample() needs to be in either output or infos...that's something you derive from the model.

Reply:
The reason why dist is not an output is that we cannot do `session.run(dist, feed_dict=fd)`.

Reply:
is `session.run(dist, feed_dict=foo)` a requirement of the API for keras.Model?